From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: "Wayne Wang(SW-TEGRA)" <waywang@nvidia.com>
Date: Tue, 29 Jul 2025 16:09:20 +0800
Subject: [PATCH 043/184] NVIDIA: SAUCE: i2c: tegra: Do not mark ACPI devices
 and PREEMPT_RT as irq safe

That patch includes commit 14d069d92951a3e150c0a81f2ca3b93e54da913b upstream.

On ACPI machines, the tegra i2c module encounters an issue due to a
mutex being called inside a spinlock. This leads to the following bug:

	BUG: sleeping function called from invalid context at kernel/locking/mutex.c:585
	...

	Call trace:
	__might_sleep
	__mutex_lock_common
	mutex_lock_nested
	acpi_subsys_runtime_resume
	rpm_resume
	tegra_i2c_xfer

The problem arises because during __pm_runtime_resume(), the spinlock
&dev->power.lock is acquired before rpm_resume() is called. Later,
rpm_resume() invokes acpi_subsys_runtime_resume(), which relies on
mutexes, triggering the error.

To address this issue, devices on ACPI are now marked as not IRQ-safe,
considering the dependency of acpi_subsys_runtime_resume() on mutexes.

Also, do not mark IRQ safe in PREEMPT_RT case as RT PM path may sleep.
And that leads to below error as RPM may not completing in time and lead
to RCU blocking.

	Call trace:
	_switch_to
	_schedule
	preempt_schedule_irq
	el1_interrupt
	el1h_64_irq_handler
	el1h_64_irq
	clk_core_disable
	clk_disable
	clk_bulk_disable
	tegra_i2c_runtime_suspend
	pm_generic_runtime_suspend

Bug 5354599
Bug 5398364

Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-jammy/+/3413225
Change-Id: Id05ae650f27495357cec1cba72b0f75fcb8befa4
Signed-off-by: Wayne Wang(SW-TEGRA) <waywang@nvidia.com>
(cherry-picked from commit a46da2899a9a0233c399d60ba5ece6e4c2978d2c)
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3419814
Reviewed-by: Sandipan Patra <spatra@nvidia.com>
Reviewed-by: Akhil R <akhilrajeev@nvidia.com>
Reviewed-by: Bibek Basu <bbasu@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
(cherry picked from commit e432fa341891cd2a230f593a20b57da257c26d86)

diff --git a/drivers/i2c/busses/i2c-tegra.c b/drivers/i2c/busses/i2c-tegra.c
index 7ee6a5c02acd..97ac0e216ee9 100644
--- a/drivers/i2c/busses/i2c-tegra.c
+++ b/drivers/i2c/busses/i2c-tegra.c
@@ -2063,8 +2063,11 @@ static int tegra_i2c_probe(struct platform_device *pdev)
 	 *
 	 * Devices with pinctrl states cannot be marked IRQ-safe as the pinctrl
 	 * state transitions during runtime PM require mutexes.
+	 *
+	 * ACPI device and RT PM case are not IRQ safe also.
 	 */
-	if (!IS_VI(i2c_dev) && !has_acpi_companion(i2c_dev->dev) && !i2c_dev->dev->pins)
+	if (!IS_VI(i2c_dev) && !has_acpi_companion(i2c_dev->dev) && !i2c_dev->dev->pins
+	    && !IS_ENABLED(CONFIG_PREEMPT_RT))
 		pm_runtime_irq_safe(i2c_dev->dev);
 
 	pm_runtime_enable(i2c_dev->dev);
-- 
2.55.0

